zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -215,44 +254,151 @@ where
+                    payment.memo.clone().unwrap_or_else(MemoBytes::empty),
+                )
+                .map_err(Error::Builder),
+            RecipientAddress::Transparent(to) => {
+                if payment.memo.is_some() {
+                    Err(Error::MemoForbidden)
+                } else {
+                    builder
+                        .add_transparent_output(&to, payment.amount)
+                        .map_err(Error::Builder)
+                }
+            }
+        }?
     }
-    .map_err(Error::Builder)?;
 
-    let consensus_branch_id = BranchId::for_height(params, height);
     let (tx, tx_metadata) = builder
         .build(consensus_branch_id, &prover)
         .map_err(Error::Builder)?;
 
-    let output_index = match to {
-        // Sapling outputs are shuffled, so we need to look up where the output ended up.
-        RecipientAddress::Shielded(_) => match tx_metadata.output_index(0) {
-            Some(idx) => idx,
-            None => panic!("Output 0 should exist in the transaction"),
-        },
-        RecipientAddress::Transparent(addr) => {
-            let script = addr.script();
-            tx.vout
-                .iter()
-                .enumerate()
-                .find(|(_, tx_out)| tx_out.script_pubkey == script)
-                .map(|(index, _)| index)
-                .expect("we sent to this address")
+    let sent_outputs = request.payments.iter().enumerate().map(|(i, payment)| {
+        let idx = match &payment.recipient_address {
+            // Sapling outputs are shuffled, so we need to look up where the output ended up.
+            RecipientAddress::Shielded(_) =>
+                tx_metadata.output_index(i).expect("An output should exist in the transaction for each shielded payment."),
+            RecipientAddress::Transparent(addr) => {
+                let script = addr.script();
+                tx.vout
+                    .iter()
+                    .enumerate()
+                    .find(|(_, tx_out)| tx_out.script_pubkey == script)
+                    .map(|(index, _)| index)
+                    .expect("An output should exist in the transaction for each transparent payment.")
+            }
+        };
+
+        SentTransactionOutput {
+            output_index: idx,
+            recipient_address: &payment.recipient_address,
+            value: payment.amount,
+            memo: payment.memo.clone()
         }
-    };
+    }).collect();
+
+    wallet_db.store_sent_tx(&SentTransaction {
+        tx: &tx,
+        created: time::OffsetDateTime::now_utc(),
+        outputs: sent_outputs,
+        account,
+        utxos_spent: vec![],
+    })
+}
+
+#[cfg(feature = "transparent-inputs")]
+#[allow(clippy::too_many_arguments)]
+pub fn shield_funds<E, N, P, D, R>(
+    wallet_db: &mut D,
+    params: &P,
+    prover: impl TxProver,
+    account: AccountId,
+    sk: &secp256k1::SecretKey,
+    extsk: &ExtendedSpendingKey,
+    memo: &MemoBytes,
+    confirmations: u32,
```

Comment:
@gmale @pacu Did we want to add a minimum shielding threshold argument here?

Reply:
Short answer, we want to be able to experiment with different confirmation limits. 

NOTE: I think there's a bug there on that function because it's adding `confirmations` to the `anchor_size`.

Reply:
When we fix this confirmations bug, we should rename the method so that callers are forced to change.